src/main/java/org/ice4j/util/PacketQueue.java

```diff
@@ -494,5 +501,332 @@ protected abstract T createPacket(
+         * to field {@link #readerFuture} and {@link #delayedFuture}.
+         */
+        private final Object syncRoot = new Object();
+
+        /**
+         * Throttle calculator which is used for counting handled packets and
+         * computing necessary delay before processing next packet when
+         * throttling is enabled.
+         */
+        private final ThrottleCalculator<T> throttler
+            = new ThrottleCalculator<>(handler);
+
+        /**
+         * Stores <tt>Future</tt> of currently executing {@link #reader}
+         */
+        private Future<?> readerFuture;
+
+        /**
+         * Stores <tt>ScheduledFuture</tt> of currently delayed
+         * execution {@link #reader}
+         */
+        private ScheduledFuture<?> delayedFuture;
+
+        /**
+         * Perpetually reads packets from this {@link PacketQueue} and uses
+         * {@link #handler} on each of them.
+         */
+        private final Runnable reader = new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                final long maxSequentiallyProcessedPackets =
+                    handler.maxSequentiallyProcessedPackets();
+
+                int sequentiallyHandledPackets = 0;
+
+                while (!closed.get())
+                {
+                    T pkt;
+
+                    synchronized (syncRoot)
+                    {
+                        long delay = throttler.getDelayNanos();
+                        if (delay > 0)
+                        {
+                            onDelay(delay, TimeUnit.NANOSECONDS);
+                            return;
+                        }
+
+
+                        if (maxSequentiallyProcessedPackets > 0 &&
+                            sequentiallyHandledPackets
+                                >= maxSequentiallyProcessedPackets)
+                        {
+                            /*
+                            All instances of AsyncPacketReader executed on
+                            single shared instance of ExecutorService to better
+                            use existing threads and to reduce number of idle
+                            threads when reader's queue is empty.
+
+                            Having limited number of threads to execute might
+                            lead to other problem, when queue is always
+                            non-empty reader will keep running, while other
+                            readers might suffer from execution starvation. One
+                            way to solve this, is to to artificially interrupt
+                            current reader execution and pump it via internal
+                            executor's queue.
+                            */
+                            onYield();
+                            return;
+                        }
+
+                        pkt = queue.poll();
```

Comment:
can you elaborate a bit on the scheme here?  i'm concerned because i tried a similar scheme in the bridge rewrite (polling queues so as to be able to shared threads) and i found it to be unusable due to the extra CPU involved in polling as opposed to just blocking on queues until there was data (with dedicated threads).  even after i played with reading them less frequently at some interval, there was always a CPU tradeoff in comparison to blocking until there was data.

Reply:
(I will say I am trying a version of poll with a 100ms timeout, but this is just to allow an opportunity for the thread to shutdown cleanly.  I may investigate a 'poison' value on the queue to achieve this and avoid the poll altogether).

Reply:
@bbaldino I have a version of this with poison value as well, it did’t push it because I though I have better chances merging simpler version :)

Reply:
but the poison still won't help you with sharing threads, no?

Reply:
> can you elaborate a bit on the scheme here? i'm concerned because i tried a similar scheme in the bridge rewrite (polling queues so as to be able to shared threads) and i found it to be unusable due to the extra CPU involved in polling as opposed to just blocking on queues until there was data (with dedicated threads). even after i played with reading them less frequently at some interval, there was always a CPU tradeoff in comparison to blocking until there was data.

I’ve just re-read this commment as well as kotlin implementation inside new libjitsi. As I can see, in a kotlin implementation there is always reschedule to poll, even if queue is currently empty. It looks like there is basically an infinite loop via recursion via scheduler queue and it will eat CPU to poll.
In my implementation I stop pollling and release thread when there is no lot items to proceed. The trigger to start polling again is a queue’s “add” method. So there must be no extra work to poll in both cases: when queue is empty and not empty, and even when adding and removing to queue is done at the same rate.

Reply:
The code pushed to github may be a bit stale, but that's a good idea (not triggering another read until the next add)!  I will experiment with that!  (Also, the call to schedule is recursive but since executor.execute just schedules the job we don't actually get an infinite recursive stack).

Reply:
> Also, the call to schedule is recursive but since executor.execute just schedules the job we don't actually get an infinite recursive stack


Yes. I got that trick :)


>  I will experiment with that!


You may try experimenting with this PR, there are variety of examples of different configuration of packet queue with different executor service, throttling and cooperative multi-tasking configuration inside benchmark tests. You could run them easily for comparison as well as introduce yours :)